src/Products/PluginIndexes/CompositeIndex/CompositeIndex.py

```diff
@@ -0,0 +1,599 @@
+                 }
+
+MIN_COMPONENTS = 2
+
+
+class ComponentMapping(PersistentMapping):
+    """A persistent wrapper for mapping objects
+    recording the order in which items are added. """
+
+    def __init__(self, *args, **kwargs):
+        self._keys = []
+        PersistentMapping.__init__(self, *args, **kwargs)
+
+    def __delitem__(self, key):
+        self._keys.remove(key)
+        PersistentMapping.__delitem__(self, key)
+
+    def __setitem__(self, key, item):
+        if key not in self._keys:
+            self._keys.append(key)
+        PersistentMapping.__setitem__(self, key, item)
+
+    def clear(self):
+        self._keys = []
+        PersistentMapping.clear(self)
+
+    def copy(self):
+        cm = ComponentMapping()
+        cm.update(self)
+        return cm
+
+    def items(self):
+        return zip(self._keys, self.values())
+
+    def keys(self):
+        return self._keys[:]
+
+    def popitem(self):
+        try:
+            key = self._keys[-1]
+        except IndexError:
+            raise KeyError('dictionary is empty')
+
+        val = self[key]
+        del self[key]
+
+        return (key, val)
+
+    def setdefault(self, key, failobj=None):
+        if key not in self._keys:
+            self._keys.append(key)
+        return PersistentMapping.setdefault(self, key, failobj)
+
+    def update(self, d):
+        for (key, val) in d.items():
+            self.__setitem__(key, val)
+
+    def values(self):
+        return map(self.get, self._keys)
+
+
+class Component(object):
+
+    _attributes = ''
+
+    def __init__(self, id, meta_type, attributes):
+        self._id = id
+        self._meta_type = meta_type
+        if attributes:
+            self._attributes = attributes
+
+    @property
+    def id(self):
+        return self._id
+
+    @property
+    def meta_type(self):
+        return self._meta_type
+
+    @property
+    def attributes(self):
+
+        attributes = self._attributes
+
+        if not attributes:
+            return [self._id]
+
+        if isinstance(attributes, str):
+            return attributes.split(',')
+
+        attributes = list(attributes)
+
+        attributes = [attr.strip() for attr in attributes if attr]
+
+        return attributes
+
+    @property
+    def rawAttributes(self):
+        return self._attributes
+
+    def __repr__(self):
+        return "<id: %s; metatype: %s; attributes: %s>" % \
+            (self.id, self.meta_type, self.attributes)
+
+
+class CompositeIndex(UnIndex):
+
+    """Index for composition of simple fields.
+       or sequences of items
+    """
+    implements(ITransposeQuery)
+
+    meta_type = "CompositeIndex"
+
+    manage_options = (
+        {'label': 'Settings',
+         'action': 'manage_main',
+         'help': ('CompositeIndex', 'CompositeIndex_Settings.stx')},
+        {'label': 'Browse',
+         'action': 'manage_browse',
+         'help': ('CompositeIndex', 'CompositeIndex_Settings.stx')},
+    )
```

Comment:
Can you remove the help part of the manage_options? HelpSys got removed with Zope 4. If there's documentation that should be available, it could go into the add or manageCompositeIndex pages, or just live in the README.